File: Examples/Ruby/benchmarking/benchmarking.rb
#!/usr/bin/env ruby

require 'json'
require 'benchmark'
require 'time'

class FlexonHelper
  def self.run_command(args)
    system("flexon-cli #{args}")
    raise "Flexon command failed" unless $?.success?
  end
end

def create_test_data
  possible_tags = ['web', 'database', 'cache', 'compute', 'storage']
  records = []

  1000.times do |i|
    records << {
      'id' => "record-#{i}",
      'timestamp' => Time.now.iso8601,
      'value' => rand * 1000,
      'metrics' => {
        'cpu' => rand * 100,
        'memory' => rand * 16384,
        'disk' => rand * 1024
      },
      'tags' => possible_tags.select { rand > 0.5 }
    }
  end

  {
    'id' => "test-#{Time.now.to_i}",
    'name' => 'Benchmark Dataset',
    'description' => 'Large dataset for benchmarking Flexon performance',
    'metadata' => {
      'created_at' => Time.now.iso8601,
      'version' => '1.0',
      'type' => 'benchmark'
    },
    'records' => records
  }
end

def benchmark_json_operations(data, iterations)
  json_ser_time = Benchmark.realtime do
    iterations.times do
      JSON.generate(data)
    end
  end

  json_str = JSON.generate(data)
  json_deser_time = Benchmark.realtime do
    iterations.times do
      JSON.parse(json_str)
    end
  end

  [json_ser_time / iterations, json_deser_time / iterations]
end

def benchmark_flexon_operations(json_file, iterations)
  ser_time = Benchmark.realtime do
    iterations.times do
      FlexonHelper.run_command("serialize -i #{json_file} -o benchmark_test.flexon")
    end
  end

  deser_time = Benchmark.realtime do
    iterations.times do
      FlexonHelper.run_command("deserialize -i benchmark_test.flexon -o benchmark_test.json")
    end
  end

  [ser_time / iterations, deser_time / iterations]
end

puts "Flexon Benchmarking Example (Ruby)"
puts "================================\n"

# Create test data
data = create_test_data
json_file = 'benchmark_data.json'
flexon_file = 'benchmark_data.flexon'
encrypted_file = 'benchmark_encrypted.flexon'

puts "1. Creating test data..."
File.write(json_file, JSON.pretty_generate(data))

# Basic serialization
puts "2. Testing basic serialization..."
FlexonHelper.run_command("serialize -i #{json_file} -o #{flexon_file}")

# Encrypted serialization
puts "3. Testing encrypted serialization..."
FlexonHelper.run_command("serialize -i #{json_file} -o #{encrypted_file} -e benchmarkkey")

# Test different encryption algorithms
puts "\n4. Testing different encryption algorithms..."

# AES-256
puts "Testing AES-256..."
FlexonHelper.run_command("serialize -i #{json_file} -o benchmark_aes.flexon -e benchmarkkey AES256")

# ChaCha20
puts "Testing ChaCha20..."
FlexonHelper.run_command("serialize -i #{json_file} -o benchmark_chacha20.flexon -e benchmarkkey ChaCha20")

# TripleDES
puts "Testing TripleDES..."
FlexonHelper.run_command("serialize -i #{json_file} -o benchmark_tripledes.flexon -e benchmarkkey TripleDES")

# Run benchmarks
puts "\n5. Running benchmarks..."
iterations = 10

puts "\nJSON Operations:"
json_ser_time, json_deser_time = benchmark_json_operations(data, iterations)
puts "Average JSON serialization time: #{json_ser_time} seconds"
puts "Average JSON deserialization time: #{json_deser_time} seconds"

puts "\nFlexon Operations:"
flexon_ser_time, flexon_deser_time = benchmark_flexon_operations(json_file, iterations)
puts "Average Flexon serialization time: #{flexon_ser_time} seconds"
puts "Average Flexon deserialization time: #{flexon_deser_time} seconds"

# Compare file sizes
puts "\nFile size comparison:"
puts "Original JSON: #{File.size(json_file)} bytes"
puts "Flexon: #{File.size(flexon_file)} bytes"
puts "AES-256: #{File.size('benchmark_aes.flexon')} bytes"
puts "ChaCha20: #{File.size('benchmark_chacha20.flexon')} bytes"
puts "TripleDES: #{File.size('benchmark_tripledes.flexon')} bytes"

# Additional benchmarks
puts "\n6. Running detailed benchmarks..."

Benchmark.bm(20) do |x|
  x.report("JSON serialize:") do
    100.times { JSON.generate(data) }
  end

  json_str = JSON.generate(data)
  x.report("JSON deserialize:") do
    100.times { JSON.parse(json_str) }
  end

  x.report("Flexon serialize:") do
    10.times do
      FlexonHelper.run_command("serialize -i #{json_file} -o benchmark_test.flexon")
    end
  end

  x.report("Flexon deserialize:") do
    10.times do
      FlexonHelper.run_command("deserialize -i benchmark_test.flexon -o benchmark_test.json")
    end
  end
end

File: Examples/PHP/benchmarking/benchmarking.php
<?php

declare(strict_types=1);

class FlexonHelper {
    public static function runCommand(string $args): void {
        $command = "flexon-cli $args";
        $result = shell_exec($command);
        if ($result === null) {
            throw new RuntimeException("Flexon command failed: $command");
        }
    }
}

function createTestData(): array {
    $possibleTags = ['web', 'database', 'cache', 'compute', 'storage'];
    $records = [];

    for ($i = 0; $i < 1000; $i++) {
        $records[] = [
            'id' => "record-$i",
            'timestamp' => date(DATE_ISO8601),
            'value' => mt_rand() / mt_getrandmax() * 1000,
            'metrics' => [
                'cpu' => mt_rand() / mt_getrandmax() * 100,
                'memory' => mt_rand() / mt_getrandmax() * 16384,
                'disk' => mt_rand() / mt_getrandmax() * 1024
            ],
            'tags' => array_filter($possibleTags, fn() => mt_rand() / mt_getrandmax() > 0.5)
        ];
    }

    return [
        'id' => "test-" . time(),
        'name' => 'Benchmark Dataset',
        'description' => 'Large dataset for benchmarking Flexon performance',
        'metadata' => [
            'created_at' => date(DATE_ISO8601),
            'version' => '1.0',
            'type' => 'benchmark'
        ],
        'records' => $records
    ];
}

function benchmarkJsonOperations(array $data, int $iterations): array {
    $jsonSerTime = microtime(true);
    for ($i = 0; $i < $iterations; $i++) {
        json_encode($data);
    }
    $jsonSerTime = (microtime(true) - $jsonSerTime) / $iterations;

    $jsonStr = json_encode($data);
    $jsonDeserTime = microtime(true);
    for ($i = 0; $i < $iterations; $i++) {
        json_decode($jsonStr, true);
    }
    $jsonDeserTime = (microtime(true) - $jsonDeserTime) / $iterations;

    return [$jsonSerTime, $jsonDeserTime];
}

function benchmarkFlexonOperations(string $jsonFile, int $iterations): array {
    $serTime = microtime(true);
    for ($i = 0; $i < $iterations; $i++) {
        FlexonHelper::runCommand("serialize -i $jsonFile -o benchmark_test.flexon");
    }
    $serTime = (microtime(true) - $serTime) / $iterations;

    $deserTime = microtime(true);
    for ($i = 0; $i < $iterations; $i++) {
        FlexonHelper::runCommand("deserialize -i benchmark_test.flexon -o benchmark_test.json");
    }
    $deserTime = (microtime(true) - $deserTime) / $iterations;

    return [$serTime, $deserTime];
}

echo "Flexon Benchmarking Example (PHP)\n";
echo "==============================\n\n";

// Create test data
$data = createTestData();
$jsonFile = 'benchmark_data.json';
$flexonFile = 'benchmark_data.flexon';
$encryptedFile = 'benchmark_encrypted.flexon';

echo "1. Creating test data...\n";
file_put_contents($jsonFile, json_encode($data, JSON_PRETTY_PRINT));

// Basic serialization
echo "2. Testing basic serialization...\n";
FlexonHelper::runCommand("serialize -i $jsonFile -o $flexonFile");

// Encrypted serialization
echo "3. Testing encrypted serialization...\n";
FlexonHelper::runCommand("serialize -i $jsonFile -o $encryptedFile -e benchmarkkey");

// Test different encryption algorithms
echo "\n4. Testing different encryption algorithms...\n";

// AES-256
echo "Testing AES-256...\n";
FlexonHelper::runCommand("serialize -i $jsonFile -o benchmark_aes.flexon -e benchmarkkey AES256");

// ChaCha20
echo "Testing ChaCha20...\n";
FlexonHelper::runCommand("serialize -i $jsonFile -o benchmark_chacha20.flexon -e benchmarkkey ChaCha20");

// TripleDES
echo "Testing TripleDES...\n";
FlexonHelper::runCommand("serialize -i $jsonFile -o benchmark_tripledes.flexon -e benchmarkkey TripleDES");

// Run benchmarks
echo "\n5. Running benchmarks...\n";
$iterations = 10;

echo "\nJSON Operations:\n";
[$jsonSerTime, $jsonDeserTime] = benchmarkJsonOperations($data, $iterations);
echo "Average JSON serialization time: $jsonSerTime seconds\n";
echo "Average JSON deserialization time: $jsonDeserTime seconds\n";

echo "\nFlexon Operations:\n";
[$flexonSerTime, $flexonDeserTime] = benchmarkFlexonOperations($jsonFile, $iterations);
echo "Average Flexon serialization time: $flexonSerTime seconds\n";
echo "Average Flexon deserialization time: $flexonDeserTime seconds\n";

// Compare file sizes
echo "\nFile size comparison:\n";
echo "Original JSON: " . filesize($jsonFile) . " bytes\n";
echo "Flexon: " . filesize($flexonFile) . " bytes\n";
echo "AES-256: " . filesize('benchmark_aes.flexon') . " bytes\n";
echo "ChaCha20: " . filesize('benchmark_chacha20.flexon') . " bytes\n";
echo "TripleDES: " . filesize('benchmark_tripledes.flexon') . " bytes\n";

// Additional benchmarks
echo "\n6. Running detailed benchmarks...\n";

echo "\nDetailed JSON Operations:\n";
$start = microtime(true);
for ($i = 0; $i < 100; $i++) {
    json_encode($data);
}
$end = microtime(true);
echo "JSON serialize (100x): " . ($end - $start) . " seconds\n";

$jsonStr = json_encode($data);
$start = microtime(true);
for ($i = 0; $i < 100; $i++) {
    json_decode($jsonStr, true);
}
$end = microtime(true);
echo "JSON deserialize (100x): " . ($end - $start) . " seconds\n";

echo "\nDetailed Flexon Operations:\n";
$start = microtime(true);
for ($i = 0; $i < 10; $i++) {
    FlexonHelper::runCommand("serialize -i $jsonFile -o benchmark_test.flexon");
}
$end = microtime(true);
echo "Flexon serialize (10x): " . ($end - $start) . " seconds\n";

$start = microtime(true);
for ($i = 0; $i < 10; $i++) {
    FlexonHelper::runCommand("deserialize -i benchmark_test.flexon -o benchmark_test.json");
}
$end = microtime(true);
echo "Flexon deserialize (10x): " . ($end - $start) . " seconds\n";
